### PR TITLE
refactor(synthetic-shadow): moving all patches as global patches

### PR DESCRIPTION
**Pull request by caridy**
## Details

* remove the global `Element.prototype.$lwcPolyfill$`, and instead we patch Element.prototype, Node.prototype, Text.prototype, HTMLElement.prototype, etc.

## Does this PR introduce breaking changes?

* 🚨 `Yes, it does introduce breaking changes.`

This is one step closer to resolve all the pending holes in DOM apis. The impact might be lower, but it is still possible that someone is using an element outside of LWC to reach inside the shadowed elements of a child lwc element.

### CODE REVIEW

File: packages/@lwc/synthetic-shadow/src/polyfills/document-body-shadow/polyfill.ts [hunk 1]
```diff
@@ -15,23 +15,12 @@ import {
     ArrayFind,
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/polyfills/document-body-shadow/polyfill.ts]: HTMLBodyElement.prototype is being patched twice, once above in faus-shadow/element.ts(https://github.com/salesforce/lwc/pull/1322/files#diff-8bc8822d07cd6e658acf12ae727b6a76R422)

Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/polyfills/document-body-shadow/polyfill.ts]: fixed

File: packages/@lwc/engine/src/__tests__/events.spec.ts [hunk 2]
```diff
@@ -485,7 +485,6 @@ describe('events', () => {
                     event = evt;
                     // div, child's shadow, x-child, parent's shadow, x-parent, body, html, document, window
                     expect(evt.composedPath().length).toBe(9);
-                    expect(evt.path.length).toBe(9);
```
Comment (caridy) [inline packages/@lwc/engine/src/__tests__/events.spec.ts]: this is not standard, so removing!

File: packages/@lwc/engine/src/__tests__/events.spec.ts [hunk 3]
```diff
@@ -518,13 +517,11 @@ describe('events', () => {
             div.click();
             // checking the value after the dispatch mechanism has finished
             const composedPath = event.composedPath();
-            const path = event.path;
-            // div, child's shadow, x-child, parent's shadow, x-parent, body, html, document, window
-            expect(composedPath.length).toBe(9);
-            expect(path.length).toBe(9);
+            // empty list
```
Comment (caridy) [inline packages/@lwc/engine/src/__tests__/events.spec.ts]: this was a bug in synthetic shadow, native always return empty composed path after the event was completely dispatched.

File: packages/@lwc/engine/src/__tests__/focus.spec.ts [hunk 4]
```diff
@@ -5,10 +5,18 @@
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
  */
 // TODO: #1311 move this tests to integration
-jest.mock('./../../env/element');
-import { isFocusable, isTabbable } from '../focus';
-
-describe('focus', () => {
+// TODO: issue #1341 fine ways to test this in browsers
```
Comment (caridy) [inline packages/@lwc/engine/src/__tests__/focus.spec.ts]: @ekashida this is the big offender... we need to find a solution for this one.

File: packages/@lwc/engine/src/__tests__/traverse.spec.ts [hunk 5]
```diff
@@ -5,10 +5,11 @@
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
  */
 import { compileTemplate } from 'test-utils';
-import { createElement, LightningElement } from '@lwc/engine';
+import { createElement, LightningElement } from '../';
 
 describe('#childNodes', () => {
-    it('should always return an empty array for slots not rendering default content', () => {
+    // TODO: issue #xxx move this test to karma, and only test this for synthetic
```
Comment (caridy) [inline packages/@lwc/engine/src/__tests__/traverse.spec.ts]: pending TODO... just need to be moved to karma.

File: packages/@lwc/engine/src/framework/__tests__/error-boundary.spec.ts [hunk 6]
```diff
@@ -101,7 +100,7 @@ describe('error boundary component', () => {
                     const boundaryHostElm = createElement('x-parent', { is: BoundaryHost });
                     document.body.appendChild(boundaryHostElm);
                     expect(
-                        querySelectorAll.call(boundaryHostElm, 'x-boundary-sibling').length
+                        boundaryHostElm.shadowRoot.querySelectorAll('x-boundary-sibling').length
```
Comment (caridy) [inline packages/@lwc/engine/src/framework/__tests__/error-boundary.spec.ts]: these tests were relying on the global unpatched API

File: packages/@lwc/engine/src/framework/hooks.ts [hunk 7]
```diff
@@ -33,12 +33,6 @@ import {
     unlockDomMutation,
     lockDomMutation,
 } from './restrictions';
-import {
```
Comment (caridy) [inline packages/@lwc/engine/src/framework/hooks.ts]: no more individual patches

File: packages/@lwc/engine/src/framework/patch.ts [hunk 8]
```diff
@@ -5,18 +5,6 @@
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
  */
 import { VNodes } from '../3rdparty/snabbdom/types';
-import { tagNameGetter } from '../env/element';
-import { setPrototypeOf, create, isUndefined } from '../shared/language';
-import { ComponentDef } from './def';
-import { HTMLElementConstructor } from './base-bridge-element';
-import { useSyntheticShadow } from './utils';
-
-// TODO: #1164 - eventually the engine should not do any of this work,
```
Comment (caridy) [inline packages/@lwc/engine/src/framework/patch.ts]: TODO: #1164

**Comment by caridy** [comment]
Folks, this is now ready to be reviewed! In the meantime I will fix the compat integration karma tests.

File: packages/@lwc/engine/src/framework/__tests__/events.spec.ts [hunk 9]
```diff
@@ -951,7 +951,7 @@ describe('Shadow Root events', () => {
             connectedCallback() {
                 this.template.addEventListener('click', evt => {
                     expect(evt.target.tagName).toBe('X-GRAND-CHILD');
-                    expect(evt.currentTarget.tagName).toBe('X-CHILD');
+                    expect(evt.currentTarget).toBe(this.template);
```
Comment (caridy) [inline packages/@lwc/engine/src/framework/__tests__/events.spec.ts]: this was also a bug in synthetic, the current target of an event listener added on the shadow must be the shadow itself. This is why we need to move these tests to karma, so we can check the behavior of the synthetic shadow.

File: packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts [hunk 10]
```diff
@@ -18,22 +14,29 @@ The complete set of contributors may be found at http://polymer.github.io/CONTRI
 Code distributed by Google as part of the polymer project is also
 subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
 */
+
+import { getOwnerDocument } from '../../shared/utils';
+import { isNull } from '../../shared/language';
+
+const GlobalShadowRoot = (window as any).ShadowRoot;
+
 export function pathComposer(startNode: EventTarget, composed: boolean): EventTarget[] {
```
Comment (caridy) [inline packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts]: this needs detailed review @ravijayaramappa

File: packages/@lwc/synthetic-shadow/src/faux-shadow/element.ts [hunk 11]
```diff
@@ -155,261 +165,299 @@ function getFirstSlottedMatch(host: HTMLElement, nodeList: NodeList): Element |
...
-        // it is a custom element, and we should then filter by slotted elements
-        return getFirstSlottedMatch(elm as HTMLElement, nodeList);
-    } else {
-        // regular element, we should then filter by ownership
-        return getFirstMatch(owner, nodeList);
+    return null;
+}
+
+function childrenGetterPatched(this: Element): HTMLCollectionOf<Element> {
+    // We cannot patch `children` in test mode
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/element.ts]: 👍 Thanks for the comments

File: packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts [hunk 12]
```diff
@@ -97,24 +97,24 @@ export function isDelegatingFocus(host: HTMLElement): boolean {
...
-export function getShadowRoot(elm: HTMLElement): SyntheticShadowRootInterface {
+export function getShadowRoot(elm: Element): SyntheticShadowRootInterface {
     return getInternalSlot(elm).shadowRoot;
 }
 
-export function hasSyntheticShadow(elm: HTMLElement): boolean {
+// Intentionally adding Node here as possible the first argument
+// since this check is harmless for nodes as well, and it speeds up things
+// to avoid casting before calling this method in few places.
+export function hasSyntheticShadow(elm: Element | Node): boolean {
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts]: How about naming this `isHostElement()`?

File: packages/@lwc/synthetic-shadow/src/faux-shadow/event-target.ts [hunk 13]
```diff
@@ -18,4 +17,73 @@ const eventCurrentTargetGetter: (this: Event) => Element | null = getOwnProperty
...
+    if (hasSyntheticShadow(this)) {
+        removeCustomElementEventListener(this, type, listener, options);
+    } else {
+        superRemoveEventListener.call(this, type, listener, options);
+    }
+}
+
+// IE11 doesn't have EventTarget, so we have to patch it conditionally:
+
+if (typeof EventTarget !== 'undefined') {
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/event-target.ts]: `if (!isUndefined(typeof EventTarget))`

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+import { getActiveElement, handleFocusIn, handleFocus, ignoreFocusIn, ignoreFocus } from './focus';
+
+const { blur } = HTMLElement.prototype;
+
+/**
+ * This method only applies to elements with a shadow attached to them
+ */
+function tabIndexGetterPatched(this: HTMLElement) {
+    if (isDelegatingFocus(this) && isFalse(hasAttribute.call(this, 'tabindex'))) {
+        // this cover the case where the default tabindex should be 0 because the
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts]: `this covers the case`

File: packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts [hunk 15]
```diff
@@ -107,74 +86,115 @@ function getFilteredSlotFlattenNodes(slot: HTMLElement): Node[] {
...
+    }
+    return parentNode as HTMLSlotElement;
 }
 
-export function PatchedSlotElement(elm: HTMLSlotElement): HTMLSlotElementConstructor {
-    const Ctor = PatchedElement(elm) as HTMLSlotElementConstructor;
-    const { addEventListener: superAddEventListener } = elm;
-    return class PatchedHTMLSlotElement extends Ctor {
-        addEventListener(
+defineProperties((window as any).HTMLSlotElement.prototype, {
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts]: Why do we sometimes access the global types off window and sometimes directly? e.g. https://github.com/salesforce/lwc/pull/1322/files#diff-6c6ff95908ccf01c5d2cf5d9a267b4d4R56
I assume it is because of the funky stuff we do with JSDOM for HTMLSlotElement.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts [hunk 16]
```diff
@@ -107,74 +86,115 @@ function getFilteredSlotFlattenNodes(slot: HTMLElement): Node[] {
...
         ) {
-            if (type === 'slotchange' && !getInternalField(this, SlotChangeKey)) {
-                if (process.env.NODE_ENV === 'test') {
-                    /* eslint-disable-next-line no-console */
-                    console.warn(
-                        'The "slotchange" event is not supported in our jest test environment.'
-                    );
-                }
+            // super.addEventListener - but that doesn't work with typescript
+            HTMLElement.prototype.addEventListener.call(this, type, listener, options);
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts]: Why not import addEventListener from ../env/element?

File: packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts [hunk 17]
```diff
@@ -18,22 +14,29 @@ The complete set of contributors may be found at http://polymer.github.io/CONTRI
...
-            (current as ShadowRoot).host &&
-            (composed || current !== startRoot)
-        ) {
-            current = (current as ShadowRoot).host as HTMLElement;
+        composedPath.push(current as Element);
+        let assignedSlot: HTMLSlotElement | null = null;
+        if (current instanceof Element) {
+            assignedSlot = current.assignedSlot;
+        }
+        if (!isNull(assignedSlot)) {
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts]:
```
        if (isTrue(current instanceof Element) && !isNull((current as Element).assignedSlot!)) {
            current = (current as Element).assignedSlot;
        }
```

File: packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts (hunk 10 above)
Comment (ravijayaramappa) [reply packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts]: Seems very similar to the original implementation if you combine the first two `if`s.

File: packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts (hunk 17 above)
Comment (ekashida) [reply packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts]: `isTrue(current instanceof Element)` seems a bit too much. The way I decide whether to use `isTrue` and other variants is if I would  have otherwise written `foo === true`. For something like`current instanceof Element`, I wouldn't have written that as `(current instanceof Element) === true` so I wouldn't use `isTrue`. Thoughts?

File: packages/@lwc/synthetic-shadow/src/env/node.ts [hunk 18]
```diff
@@ -22,9 +22,26 @@ const {
...
+    Node.prototype,
+    'lastChild'
+)!.get!;
+
+const textContentGetter: (this: Node) => string = getOwnPropertyDescriptor(
+    Node.prototype,
+    'textContent'
+)!.get!;
+
+const parentNodeGetter: (this: Node) => (Node & ParentNode) | null = getOwnPropertyDescriptor(
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/env/node.ts]: `(Node & ParentNode)` looks redundant, is the `Node` bit necessary?

File: packages/@lwc/engine/src/framework/__tests__/events.spec.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/engine/src/framework/__tests__/events.spec.ts]: I haven't fixed the bug in synthetic shadow just yet, will do that once we have this test is karma and it start failing.

File: packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts [hunk 19]
```diff
@@ -18,22 +14,29 @@ The complete set of contributors may be found at http://polymer.github.io/CONTRI
...
-            current = (current as HTMLElement).assignedSlot as HTMLSlotElement;
-        } else if (
-            (current as HTMLElement).nodeType === DOCUMENT_FRAGMENT_NODE &&
-            (current as ShadowRoot).host &&
-            (composed || current !== startRoot)
-        ) {
-            current = (current as ShadowRoot).host as HTMLElement;
+        composedPath.push(current as Element);
+        let assignedSlot: HTMLSlotElement | null = null;
+        if (current instanceof Element) {
```
[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts]: yes, but Text is not an EventTarget, nor a ParentNode, so, it is safe. I will add a comment though.

File: packages/@lwc/synthetic-shadow/src/env/mutation-observer.ts [hunk 20]
```diff
@@ -5,32 +5,9 @@
...
-}
-
-const MutationObserver = MO;
-
 // Eventually, import the patched MutationObserver polyfill here
 // to ensure rest of the framework uses the patched version
 
-const MutationObserverObserve = MutationObserver.prototype.observe;
-export { MutationObserver, MutationObserverObserve };
+const MO = MutationObserver;
```
[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/env/mutation-observer.ts]: sure, there is already a comment, but I will expand it.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts (hunk 12 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts]: good call... will do

File: packages/@lwc/synthetic-shadow/src/faux-shadow/event-target.ts (hunk 13 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/event-target.ts]: It is not the same... typeof always return string values, while isUndefined() tests for === with `undefined`

File: packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts (hunk 16 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts]: it is a little bit more tricky, maybe we can accommodate that in another refactor?

File: packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts (hunk 15 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts]: it is not longer about JSDOM, but it is about typescript... in this case it is fine to remove that... will do...

File: packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts (hunk 17 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts]: good q! maybe we should open an issue to track that since it is a larger change through multiple pkgs.

File: packages/@lwc/synthetic-shadow/src/env/node.ts (hunk 18 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/env/node.ts]: that's to match TS definitions so it doesn't complain about defining dom APIs with less general types.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/element.ts [hunk 21]
```diff
@@ -155,261 +165,299 @@ function getFirstSlottedMatch(host: HTMLElement, nodeList: NodeList): Element |
...
-            this: EventTarget,
-            type: string,
-            listener: EventListener,
-            options?: boolean | AddEventListenerOptions
-        ) {
-            if (hasSyntheticShadow(this as HTMLElement)) {
-                removeCustomElementEventListener(this as HTMLElement, type, listener, options);
-            } else {
-                superRemoveEventListener.call(this as HTMLElement, type, listener, options);
+            // TODO: make this a global patch with a way to disable it
```
[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/element.ts]: @ravijayaramappa @ekashida this is actually important... we have few more global patches that we don't have in place today... and I'm still questioning myself if our current escape hatch should apply to those apis (similar to what we do with querySelectors and others when executed outside of the scope of LWC), the apis in question are:

Note: cloneNode, contains, compareDocumentPosition, textContent
Element: outerHTML, innerHTML

I think we can probably add it to cloneNode, contains, compareDocumentPosition, but the other 3 seems like not a problem, unless you're doing something very wrong at the aura level... to get the content and test it or something, what do you think?

/cc @trevor-bliss in case he knows.

File: packages/@lwc/engine/src/__tests__/dom.spec.ts [hunk 22]
```diff
@@ -18,8 +18,8 @@ describe('dom', () => {
             });
             elm.click();
         });
-        // Flapper
```
Comment (ekashida) [inline packages/@lwc/engine/src/__tests__/dom.spec.ts]: I recently saw this in another PR and didn't understand how such a simple test could flap. Did we change something to address that?

Comment (caridy) [reply packages/@lwc/engine/src/__tests__/dom.spec.ts]: well, we are NOT longer using synthetic shadow in jest, will that help? I think it does.

File: packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (ravijayaramappa) [reply packages/@lwc/synthetic-shadow/src/3rdparty/polymer/path-composer.ts]: Text extends EventTarget.
Screen Shot 2019-06-06 at 1 17 36 PM
https://developer.mozilla.org/en-US/docs/Web/API/Text

File: packages/@lwc/synthetic-shadow/src/faux-shadow/element.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (ravijayaramappa) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/element.ts]: I feel like we should have the escape hatch for these APIs. They are high traffic and in the likely hood of a failures, the escape hatch gives us some wiggle room.
```Node: cloneNode, textContent
Element: outerHTML, innerHTML```

[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/element.ts]: ✅

File: packages/@lwc/synthetic-shadow/src/env/element.ts [hunk 23]
```diff
@@ -19,36 +19,52 @@ const {
...
+
 const matches: (this: Element, selector: string) => boolean = hasOwnProperty.call(
     Element.prototype,
     'matches'
 )
     ? Element.prototype.matches
     : (Element.prototype as any).msMatchesSelector; // IE11
 
-const childrenGetter: (this: HTMLElement) => HTMLCollectionOf<Element> = hasOwnProperty.call(
+const childrenGetter: (this: Element) => HTMLCollectionOf<Element> = hasOwnProperty.call(
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/env/element.ts]:
```suggestion
const childrenGetter: (this: ParentNode) => HTMLCollectionOf<Element> = hasOwnProperty.call(
```

File: packages/@lwc/synthetic-shadow/src/env/element.ts [hunk 24]
```diff
@@ -19,36 +19,52 @@ const {
...
+const firstElementChildGetter: (this: Element) => Element | null = getOwnPropertyDescriptor(
     Element.prototype,
-    'outerHTML'
-)
-    ? getOwnPropertyDescriptor(Element.prototype, 'outerHTML')!.set!
-    : getOwnPropertyDescriptor(HTMLElement.prototype, 'outerHTML')!.set!; // IE11
+    'firstElementChild'
+)!.get!;
+
+const lastElementChildGetter: (this: Element) => Element | null = getOwnPropertyDescriptor(
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/env/element.ts]:
```suggestion
const lastElementChildGetter: (this: ParentNode) => Element | null = getOwnPropertyDescriptor(
```

File: packages/@lwc/synthetic-shadow/src/env/element.ts [hunk 25]
```diff
@@ -19,36 +19,52 @@ const {
...
     Element.prototype,
-    'innerHTML'
-)
-    ? getOwnPropertyDescriptor(Element.prototype, 'innerHTML')!.set!
-    : getOwnPropertyDescriptor(HTMLElement.prototype, 'innerHTML')!.set!; // IE11
+    'childElementCount'
+)!.get!;
 
-const outerHTMLSetter: (this: Element, s: string) => void = hasOwnProperty.call(
+const firstElementChildGetter: (this: Element) => Element | null = getOwnPropertyDescriptor(
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/env/element.ts]:
```suggestion
const firstElementChildGetter: (this: ParentNode) => Element | null = getOwnPropertyDescriptor(
```

File: packages/@lwc/synthetic-shadow/src/env/element.ts [hunk 26]
```diff
@@ -19,36 +19,52 @@ const {
 
 const { addEventListener, removeEventListener } = Element.prototype;
 
-const innerHTMLSetter: (this: Element, s: string) => void = hasOwnProperty.call(
+const childElementCountGetter: (this: Element) => number = getOwnPropertyDescriptor(
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/env/element.ts]:
```suggestion
const childElementCountGetter: (this: ParentNode) => number = getOwnPropertyDescriptor(
```

File: packages/@lwc/synthetic-shadow/src/faux-shadow/event-target.ts [hunk 27]
```diff
@@ -18,4 +17,73 @@ const eventCurrentTargetGetter: (this: Event) => Element | null = getOwnProperty
     'currentTarget'
 )!.get!;
 
+// These methods are usually from EventTarget.prototype, but that's not available in IE11, the best best thing
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/event-target.ts]:
```suggestion
// These methods are usually from EventTarget.prototype, but that's not available in IE11, the next best thing
```

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts [hunk 28]
```diff
@@ -338,24 +332,17 @@ function isValidEventForCustomElement(event: Event): boolean {
 }
 
 export function addCustomElementEventListener(
-    elm: HTMLElement,
+    elm: Element,
     type: string,
     listener: EventListener,
-    options?: boolean | AddEventListenerOptions
+    _options?: boolean | AddEventListenerOptions
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts]: Was this change necessary or does the underscore imply something? Is it because we're not using this argument but we need it for typescript?

File: packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts [hunk 29]
```diff
@@ -18,24 +18,24 @@ import {
...
     getNodeOwner,
     getAllMatches,
     getFilteredChildNodes,
-} from './traverse';
-import { childNodesGetter as nativeChildNodesGetter } from '../env/node';
+    getFilteredSlotAssignedNodes,
+} from '../faux-shadow/traverse';
+import {
+    childNodesGetter as nativeChildNodesGetter,
+    childNodesGetter,
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts]: Looks like we're importing this twice.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/text.ts [hunk 30]
```diff
@@ -0,0 +1,18 @@
...
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { defineProperties } from '../shared/language';
+import { assignedSlotGetterPatched } from './slot';
+
+// Non-deep-traversing patches
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/text.ts]: I didn't understand this comment.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts (hunk 28 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts]: yes, it is exactly that. we have that rule in the ts config.

File: packages/@lwc/synthetic-shadow/README.md [hunk 31]
```diff
@@ -6,7 +6,7 @@ This is a polyfill for ShadowRoot that was tailor-made for LWC in order to meet
 
 ## Compromises
 
-TBD
+-   Default content for `<slot>` elements is always empty.
```
[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/synthetic-shadow/README.md]: yes!

[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/synthetic-shadow/README.md]: added

File: packages/@lwc/engine/src/shared/assert.ts [hunk 32]
```diff
@@ -6,41 +6,30 @@
...
+        );
+        indentation = indentation + indentationChar;
+        const newRootNode = currentNode.getRootNode();
+        if (newRootNode === currentNode || newRootNode === document) {
+            currentNode = null; // quitting
+        } else if (newRootNode instanceof ShadowRoot) {
+            currentNode = newRootNode.host;
         } else {
-            currentElement = parentNodeGetter.call(currentElement);
+            currentNode = newRootNode;
```
[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/engine/src/shared/assert.ts]: will add the comment.
